### DISCUSSION TITLE
Why does TakenAt get overridden by CreatedAt when indexing?

### DISCUSSION THREAD

**Discussion by mikamikem** [photoprism/photoprism#3282] [Q&A] [answered]
Hello!

Thanks for making PhotoPrism, I'm just getting started with it, but so far it seems incredibly useful and I'm excited to get it fully up and running with all my photos soon!

### My Specific Context:
I keep forgetting to set the date and time on my camera and have a number of pictures with an incorrect date and time set by the camera.  I update the date and time in Lightroom Classic which updates the EXIF field `DateTimeOriginal` to the correct values and I then export the picture with the updated timestamp to a location that's indexed by PhotoPrism.  I do not export or generate an XMP or any sidecar file, so the indexer should be relying entirely on the exported jpeg.

When I index the pictures with corrected date and times, it shows the date and time originally set by the camera (which is still present in the EXIF and XMP fields that exiftool calls `CreateDate`, but which might be `DateTimeDigitized`) instead of the one that I updated in Lightroom Classic and that exiftool shows in the `DateTimeOriginal` field.

### Technical Background and General Context:
In commit 3403c50, there [was a check added](https://github.com/photoprism/photoprism/commit/3403c50c48de7c1cecadaf6b822d9f35d99e4b3e#diff-5e54ffefc2cd1c77d57258d92f2c45a9d9cebf1ad8b70519309cd32eae287de1R197) that says if the `CreatedAt` meta data values were set, that they should overwrite the `TakenAt` field.  Shortly after, there was [another check added](https://github.com/photoprism/photoprism/commit/3403c50c48de7c1cecadaf6b822d9f35d99e4b3e#diff-5e54ffefc2cd1c77d57258d92f2c45a9d9cebf1ad8b70519309cd32eae287de1R210) so that if the value of `TakenAtLocal` does not match the `TakenAt` value (which is now the `CreatedAt` value if any of those meta keys exist) by more than 27 hours, it replaces the `TakenAtLocal` with the `TakenAt` value (which again may have been changed to `CreatedAt`).

In my workflow and from what I can gather looking at the code, my camera sets one of the `CreatedAt` keys with an inaccurate date and time when it generates the original source image (so it always replaces `TakenAt`) and the value is years in the past which causes it to fail the `TakenAt` to `TakenAtLocal` comparison check (and so it also replaces `TakenAtLocal` with `CreatedAt`'s value).  This means that PhotoPrism effectively ignores my change to `DateTimeOriginal` that I make in Lightroom Classic entirely.

I've seen a lot of people describe using the workflow I followed to fix up incorrectly timestamped photos (as well as setting a proper `DateTimeOriginal` for scanned photos) and the only other recommendations were to manually adjust the originals with exiftool...which seems less than ideal for an average user.  From what I can find about the spec, it seems like the `DateTimeOriginal` field is a bit ambiguous, but it seems like people generally agree that it's supposed to represent the time that the picture being represented was taken, whereas the `DateTimeDigitized` (which is what exiftool reports as `CreateDate` and that I think is parsed into `CreatedAt` by PhotoPrism, although I'm still not 100% sure on that part since it's also listed under `TakenAt`) is, as it says, the time when the image was digitized.  In cases where a photo is scanned, these can vary quite a lot as a normal and expected way for those fields to work (without me needing to mess up my camera's date and time settings and forgetting to fix them before I start taking photos).

While I guess you could make the argument that you may want to sort those scans (or fixed up photos) by digitized date in PhotoPrism, it seems more logical in my mind to want to store them by when the original photo was taken.  Like if I had a childhood photo that was scanned, I would want it to be sorted in all my pictures by when it was taken and not by when I happened to scan it.

### Questions:
1. Am I reading the code right, that the value stored in `DateTimeOriginal` would be replaced for sorting purposes in PhotoPrism by any metadata value represented by `CreatedAt` (if any of the keys were set)?

If so:

2. Why do you prefer the value in `CreatedAt` over the value in `TakenAt`?
3. I believe the table at https://www.photoprism.app/kb/metadata is generated from code, but if it's possible to add something that indicates the values in `TakenAt` (and possibly also `TakenAtLocal`) are superseded by the values in `CreatedAt`, that would be incredibly helpful.  I was looking at that table a lot in my misadventures trying to sort this out and thinking that I had gone crazy before I started just looking at the code and found the checks that override the values.

Thanks for any thoughts on all of this!  I'm excited to get more pictures imported to see how it works on a bigger collection and hopefully be able to contribute back to the project!

Thanks again!
Nora

### Addendum

My file's output from `exiftool -G0:1:2 -j file.jpg | grep Date`

  "File:System:Time:FileModifyDate": "2023:03:13 15:57:40-07:00",
  "File:System:Time:FileAccessDate": "2023:03:13 15:58:02-07:00",
  "File:System:Time:FileInodeChangeDate": "2023:03:13 15:57:40-07:00",
  "EXIF:IFD0:Time:ModifyDate": "2023:03:13 15:57:38",
  "EXIF:ExifIFD:Time:DateTimeOriginal": "2023:03:11 10:38:52",
  "EXIF:ExifIFD:Time:CreateDate": "2009:02:05 13:25:30",
  "IPTC:Time:DateCreated": "2023:03:11",
  "IPTC:Time:DigitalCreationDate": "2009:02:05",
  "ICC_Profile:ICC-header:Time:ProfileDateTime": "1998:02:09 06:49:00",
  "XMP:XMP-xmp:Time:ModifyDate": "2023:03:13 15:57:38-07:00",
  "XMP:XMP-xmp:Time:CreateDate": "2009:02:05 13:25:30",
  "XMP:XMP-xmp:Time:MetadataDate": "2023:03:13 15:57:38-07:00",
  "XMP:XMP-photoshop:Time:DateCreated": "2023:03:11 10:38:52",
  "Composite:Time:SubSecModifyDate": "2023:03:13 15:57:38-07:00",
  "Composite:Time:DateTimeCreated": "2023:03:11 10:38:52",
  "Composite:Time:DigitalCreationDateTime": "2009:02:05 13:25:30",

And in PhotoPrism, I see the date February 5, 2009 with local time set to 13:25:30.

**Comment by graciousgrey** [comment, marked as answer]
For images from digital cameras, CreateDate and DateTimeOriginal should be approximately the same except for the time zone. CreateDate is usually in UTC and DateTimeOriginal is in local time.

When CreateDate and DateTimeOriginal are present, PhotoPrism uses the CreateDate because UTC time is more accurate. Especially if the image has no coordinates.

We understand that scanners can use the fields differently, but we don't have the resources right now to optimize PhotoPrism for this.
 
One solution would be to remove the CreateDate from the metadata.

Adding a hin to our docs is a good idea :+1:

**Reply by mikamikem** [reply]
Thanks for the quick reply!  That makes sense if the norm for digital cameras is to use UTC for `CreateDate` and local for `DateTimeOriginal`.  I couldn't find any reliable info on that when I was searching before my original post, so it's good to know that's the most common scenario and my camera is just weird and/or old.  While the timezone on my camera is defaulted to the wrong thing as well, it is set to something that isn't UTC or GMT, so apparently my camera saves out local time to all the fields instead of using UTC for `CreateDate`.

I've worked around the issue by dropping the bad metadata as you suggested (see below), but I am still curious why the large mismatch is assumed to be bad data in `TakenAt` and not bad data in `CreatedAt` or even intentionally different.  If `CreatedAt` and `TakenAt` are drastically different, it feels like defaulting to `CreatedAt` is an arbitrary choice over `TakenAt`, especially when things like Lightroom only allow modifying `TakenAt` and the scanned files workflow would have very different meanings for those two values.  It seems like the logic in PhotoPrism could stay the same except for the offset mismatch check, which could revert `TakenAt` to it's original value if the offset is too large.  That seems like it would solve the issue for scanned photos and for a lot of manually fixed photos (exception below) without disrupting the existing workflow where `CreatedAt` and `TakenAt` have valid values in UTC and local time as you described.  I'm assuming there's a more common use case for overriding the value even when there's a large mismatch?

It also seems like there's a silent assumption in the existing code that could be misleading where the code that determines if local time should be overwritten assumes that a less than 27 hour time difference in the date and time is attributed to the difference between UTC and local time.  There's a very common case where that would not be accurate with the Lightroom workflow again, where people do quick modifications to offset the time by a timezone difference (again because they forgot to update it manually when they were traveling).

If I'm following the code properly, setting your time properly before you leave, forgetting to update while traveling to another time zone, and updating it in Lightroom when you get back would cause PhotoPrism to interpret the offset between your local time and your travel destination's time as the time offset from UTC.  Again since the fields are not well defined or consistently used, this may be impossible to detect automatically without also breaking the ability to detect the time offset automatically for cameras that use UTC in `CreateDate` and local in `DateTimeOriginal`, but it is unexpected behavior and currently wouldn't generate even a warning in the logs.

### Workaround
Anyway, for my original issue, not sure how many people use Lightroom Classic on here, but I came up with a workaround to mostly automatically drop the metadata as you suggested for files that I changed the date and time on.  I used Jeffrey Friedl's [Metadata Wrangler plugin](http://regex.info/blog/lightroom-goodies/metadata-wrangler) for Lightroom Classic to remove the tags automatically as part of the publish step (also works for export).  If you want to set it up, I did the following in Lightroom Classic to fix this for my pictures:

1. Tagged my photos that needed the time fixed with a specific keyword
2. Made a new publish service (I'm using his Folder Publisher plugin to do the publish itself) for the time fixup files
3. Set up a smart collection filtering for the tag from step 1 (and whatever my normal criteria were) under the new publish service
5. Activated the Metadata Wrangler post-process action and put the following in the field list:
```
EXIF:ExifIFD:Time:CreateDate,IPTC:Time:DigitalCreationDate,IPTC:Time:DigitalCreationTime,XMP:XMP-xmp:Time:CreateDate,Composite:Time:DigitalCreationDateTime
```
6. Changed my original smart collection to not include anything with the tag I made in step 1 so that I only clear the metadata from pictures that have bad data.

Now I have two publish services that automatically update if I make sure to tag the photos appropriately when I manually update the timestamps.  The photos with the metadata removed sort appropriately in PhotoPrism now.



Thanks again for the quick info!
Nora

**Reply by ndfan77** [reply]
This should be optional or configurable behavior.  When a good portion of images come from scanner software, which sets the exif CreateDate to when the image was acanned, we should be able to set the TakenAt date and have it be recognized before CreateDate, so that we don't have to erase CreateDate and lose when the image was acanned.
